### PR TITLE
Provide the correct address of the programs table

### PR DESCRIPTION
**Pull request by kasperl**
Now communicated through the image info blob, instead of relying on weird link table entries.

### CODE REVIEW

File: src/os_esp32.cc [hunk 1]
```diff
@@ -741,19 +741,20 @@ void OS::heap_summary_report(int max_pages, const char* marker) { }
 #endif // def TOIT_CMPCTMALLOC
 
 static const int TOIT_IMAGE_DATA_SIZE = 1024;
-static const int TOIT_CONFIG_IMAGE_SIZE = TOIT_IMAGE_DATA_SIZE - UUID_SIZE;
+static const int TOIT_CONFIG_IMAGE_SIZE = TOIT_IMAGE_DATA_SIZE - UUID_SIZE - sizeof(uint32);
```
Comment (floitsch) [inline src/os_esp32.cc:744]: Make that a constant.
Also explain why they are removed. (I'm guessing we reserve this space).

File: tools/firmware.toit [hunk 2]
```diff
@@ -726,15 +727,22 @@ inject_config config/Map unique_id/uuid.Uuid bits/ByteArray -> ByteArray:
   image_data_position := find_image_data_position bits
   image_data_offset := image_data_position[0]
   image_data_size := image_data_position[1]
-  image_config_size := image_data_size - uuid.SIZE
+  image_config_size := image_data_size - uuid.SIZE - 4
```
Comment (floitsch) [inline tools/firmware.toit:730]: Use a constant and/or explain the "4".

File: src/resources/image.cc [hunk 3]
```diff
@@ -99,11 +99,11 @@ PRIMITIVE(writer_commit) {
...
   // with the program image. Check that the size of the encoded assets is within
   // bounds and mark the metadata to indicate the presence of the assets.
-  int program_size = output->program_size();
-  int assets_extra = image.byte_size() - program_size;
+  uword program_size = output->program_size();
+  uword assets_extra = image.byte_size() - program_size;
   if (assets_extra > 0) {
     uword assets_address = reinterpret_cast<uword>(image.begin()) + program_size;
-    int assets_length = *reinterpret_cast<uint32*>(assets_address);
+    uword assets_length = *reinterpret_cast<uint32*>(assets_address);
```
Comment (floitsch) [inline src/resources/image.cc:106]:
```suggestion
    uint32 assets_length = *reinterpret_cast<uint32*>(assets_address);
```
or 
```suggestion
    uword assets_length = *reinterpret_cast<uword*>(assets_address);
```

I'm guessing it's encoded as 32bit integer, but I didn't check.

File: tools/firmware.toit [hunk 4]
```diff
@@ -726,15 +727,22 @@ inject_config config/Map unique_id/uuid.Uuid bits/ByteArray -> ByteArray:
...
+  // Determine the address of the bundled programs table by decoding
+  // the segments of the binary.
+  binary := Esp32Binary bits
+  bundled_programs_table_address := ByteArray 4
+  LITTLE_ENDIAN.put_uint32 bundled_programs_table_address 0 binary.extend_drom_address
+
   bits.replace image_data_offset (ByteArray image_data_size)  // Zero out area.
   bits.replace image_data_offset config_data
   bits.replace image_data_offset + image_config_size unique_id.to_byte_array
+  bits.replace image_data_offset + image_data_size - 4 bundled_programs_table_address
```
Comment (floitsch) [inline tools/firmware.toit:745]:
```suggestion
  bits.replace (image_data_offset + image_config_size) unique_id.to_byte_array
  bits.replace (image_data_offset + image_data_size - 4) bundled_programs_table_address
```

**Review by erikcorry — APPROVED** [review]

**Review by floitsch — APPROVED** [review]
LGTM.

File: tools/firmware.toit (hunk 2 above)
Comment (kasperl) [reply tools/firmware.toit:730]: Will get rid of this code completely.

File: src/resources/image.cc (hunk 3 above)
Comment (kasperl) [reply src/resources/image.cc:106]: It is a 32-bit integer.

**Pull request merged**